File: diagnose-local.py

#!/usr/bin/env python3
"""
Diagnostic script for local EuroMillions database
Python version of diagnose-local.js
"""

import json
import urllib.request
import urllib.error
from datetime import datetime
import sys
import platform

BASE_URL = 'http://localhost:5000'

def colored_print(message, color='white'):
    """Print colored text to console"""
    colors = {
        'green': '\033[92m',
        'red': '\033[91m',
        'yellow': '\033[93m',
        'blue': '\033[94m',
        'white': '\033[97m',
        'reset': '\033[0m'
    }
    print(f"{colors.get(color, colors['white'])}{message}{colors['reset']}")

def make_request(url, timeout=5):
    """Make HTTP request and return response"""
    try:
        with urllib.request.urlopen(url, timeout=timeout) as response:
            data = response.read().decode('utf-8')
            return {
                'status': response.status,
                'success': response.status == 200,
                'data': json.loads(data),
                'error': None
            }
    except urllib.error.HTTPError as e:
        return {
            'status': e.code,
            'success': False,
            'data': None,
            'error': f'HTTP {e.code}: {e.reason}'
        }
    except urllib.error.URLError as e:
        error_msg = str(e.reason)
        if 'Connection refused' in error_msg:
            error_msg = 'Connection refused - server is not running'
        elif 'Name or service not known' in error_msg:
            error_msg = 'Server not found - check if localhost:5000 is accessible'
        return {
            'status': 0,
            'success': False,
            'data': None,
            'error': error_msg
        }
    except json.JSONDecodeError as e:
        return {
            'status': 0,
            'success': False,
            'data': None,
            'error': f'Invalid JSON: {e}'
        }
    except Exception as e:
        return {
            'status': 0,
            'success': False,
            'data': None,
            'error': f'Request timeout' if 'timeout' in str(e) else f'Unexpected error: {e}'
        }

def main():
    colored_print('🔍 EuroMillions Database Diagnostic Tool', 'blue')
    colored_print('', 'blue')
    colored_print(f'📍 Testing connection to: {BASE_URL}', 'yellow')
    
    # Check if server is running
    colored_print('\n1. Checking server connection...', 'blue')
    health_check = make_request(f'{BASE_URL}/api/stats')
    
    if not health_check['success']:
        colored_print('❌ Server not responding. Make sure it\'s running with "python run-local.py"', 'red')
        colored_print(f'   Error: {health_check["error"] or "Unknown error"}', 'red')
        
        if health_check['error'] and ('Connection refused' in health_check['error'] or 'not found' in health_check['error']):
            colored_print('\n💡 Environment Detection:', 'yellow')
            colored_print(f'   You are running this diagnostic on your local {platform.system()} machine,', 'yellow')
            colored_print('   but there\'s no server running locally.', 'yellow')
            colored_print('\n🔧 Choose one solution:', 'yellow')
            colored_print('   Option A (Recommended): Run this diagnostic on Replit where the server is running', 'yellow')
            colored_print('   Option B: Start the server locally on your machine:', 'yellow')
            colored_print('     1. Open a NEW terminal window (keep this one open)', 'yellow')
            colored_print('     2. Navigate to your project directory', 'yellow')
            colored_print('     3. Run: python run-local.py', 'yellow')
            colored_print('     4. Wait for "serving on localhost:5000" message', 'yellow')
            colored_print('     5. Then run this diagnostic again: python diagnose-local.py', 'yellow')
            colored_print('\n   📋 Before running the server, ensure:', 'yellow')
            colored_print('     - Your .env file has a valid DATABASE_URL', 'yellow')
            colored_print('     - Node.js 18+ is installed', 'yellow')
            colored_print('     - You have a PostgreSQL database setup', 'yellow')
            colored_print('\n   ✅ Status: The server is running successfully on Replit with 50 draws loaded.', 'yellow')
        else:
            colored_print('   This appears to be a network connectivity issue.', 'yellow')
            colored_print('   The server you\'re trying to reach (localhost:5000) is not responding.', 'yellow')
            colored_print('   Make sure you have a local server running or use the Replit version.', 'yellow')
        return
    
    colored_print('✅ Server is running', 'green')
    
    # Check database stats
    colored_print('\n2. Checking database stats...', 'blue')
    stats = make_request(f'{BASE_URL}/api/stats')
    
    if not stats['success']:
        colored_print('❌ Stats API failed', 'red')
        colored_print(f'   Status: {stats["status"]}', 'red')
        colored_print(f'   Error: {stats["error"] or "Unknown error"}', 'red')
    else:
        colored_print('✅ Database stats loaded', 'green')
        data = stats['data']
        colored_print(f'   Total combinations: {data["totalCombinations"]:,}', 'blue')
        colored_print(f'   Drawn combinations: {data["drawnCombinations"]}', 'blue')
        colored_print(f'   Never drawn: {data["neverDrawnCombinations"]:,}', 'blue')
    
    # Check history
    colored_print('\n3. Checking draw history...', 'blue')
    history = make_request(f'{BASE_URL}/api/history')
    
    if not history['success']:
        colored_print('❌ History API failed', 'red')
        colored_print(f'   Status: {history["status"]}', 'red')
        colored_print(f'   Error: {history["error"] or "Unknown error"}', 'red')
    else:
        colored_print('✅ Draw history loaded', 'green')
        data = history['data']
        colored_print(f'   Number of draws: {len(data)}', 'blue')
        if len(data) > 0:
            latest = data[0]
            draw_date = datetime.fromisoformat(latest['drawDate'].replace('Z', '+00:00'))
            colored_print(f'   Latest draw: {draw_date.strftime("%Y-%m-%d")}', 'blue')
            colored_print(f'   Numbers: {", ".join(map(str, latest["numbers"]))}', 'blue')
            colored_print(f'   Stars: {", ".join(map(str, latest["stars"]))}', 'blue')
    
    # Force initialization
    colored_print('\n4. Force initializing data...', 'blue')
    init = make_request(f'{BASE_URL}/api/initialize')
    
    if not init['success']:
        colored_print('❌ Initialization failed', 'red')
        colored_print(f'   Status: {init["status"]}', 'red')
        colored_print(f'   Error: {init["error"] or "Unknown error"}', 'red')
    else:
        colored_print('✅ Data initialization completed', 'green')
        if init['data'] and 'stats' in init['data']:
            stats_data = init['data']['stats']
            colored_print(f'   Draws loaded: {stats_data["drawnCombinations"]}', 'blue')
            colored_print(f'   Prediction accuracy: {stats_data["predictionAccuracy"] * 100:.1f}%', 'blue')
    
    # Check analytics
    colored_print('\n5. Checking analytics after initialization...', 'blue')
    analytics = make_request(f'{BASE_URL}/api/analytics/numbers')
    
    if not analytics['success']:
        colored_print('❌ Analytics API failed', 'red')
        colored_print(f'   Status: {analytics["status"]}', 'red')
        colored_print(f'   Error: {analytics["error"] or "Unknown error"}', 'red')
    else:
        colored_print('✅ Analytics loaded', 'green')
        data = analytics['data']
        if 'hotNumbers' in data and data['hotNumbers']:
            hot_numbers = [str(n['number']) for n in data['hotNumbers'][:5]]
            colored_print(f'   Hot numbers: {", ".join(hot_numbers)}', 'blue')
        if 'hotStars' in data and data['hotStars']:
            hot_stars = [str(s['number']) for s in data['hotStars'][:3]]
            colored_print(f'   Hot stars: {", ".join(hot_stars)}', 'blue')
    
    # Final summary
    colored_print('\n📋 Diagnostic Summary:', 'blue')
    
    all_working = (health_check['success'] and stats['success'] and 
                   history['success'] and init['success'] and analytics['success'])
    
    if all_working:
        colored_print('✅ All systems working correctly!', 'green')
        colored_print('   Your local setup is ready to use at http://localhost:5000', 'green')
    else:
        colored_print('⚠️  Some issues detected. Recommendations:', 'yellow')
        
        if not (stats['success'] and history['success'] and analytics['success']):
            colored_print('   1. Check your DATABASE_URL in .env file', 'yellow')
            colored_print('   2. Run: npm run db:push', 'yellow')
            colored_print('   3. Restart the server: python run-local.py', 'yellow')
            colored_print('   4. Visit: http://localhost:5000/api/initialize', 'yellow')
    
    colored_print('\n🏁 Diagnostic complete.\n', 'blue')

if __name__ == "__main__":
    main()

File: diagnose-local.js

/**
 * Diagnostic script for local EuroMillions database
 * Run with: node diagnose-local.js
 */

import http from 'http';
const BASE_URL = 'http://localhost:5000';

// Colors for console output
const colors = {
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  reset: '\x1b[0m'
};

function log(message, color = 'reset') {
  console.log(colors[color] + message + colors.reset);
}

async function makeRequest(url) {
  return new Promise((resolve) => {
    const req = http.get(url, (res) => {
      let data = '';
      
      res.on('data', (chunk) => {
        data += chunk;
      });
      
      res.on('end', () => {
        try {
          const json = JSON.parse(data);
          resolve({
            status: res.statusCode,
            success: res.statusCode === 200,
            data: json,
            error: null
          });
        } catch (e) {
          resolve({
            status: res.statusCode,
            success: false,
            data: data,
            error: 'Invalid JSON: ' + e.message
          });
        }
      });
    });
    
    req.on('error', (error) => {
      let errorMessage = error.message;
      
      if (error.code === 'ECONNREFUSED') {
        errorMessage = 'Connection refused - server is not running';
      } else if (error.code === 'ENOTFOUND') {
        errorMessage = 'Server not found - check if localhost:5000 is accessible';
      } else if (error.code === 'ETIMEDOUT') {
        errorMessage = 'Connection timeout - server may be starting up';
      }
      
      resolve({
        status: 0,
        success: false,
        data: null,
        error: errorMessage,
        code: error.code
      });
    });
    
    req.setTimeout(5000, () => {
      req.destroy();
      resolve({
        status: 0,
        success: false,
        data: null,
        error: 'Request timeout'
      });
    });
  });
}

async function diagnose() {
  log('🔍 EuroMillions Database Diagnostic Tool\n', 'blue');
  log('📍 Testing connection to: ' + BASE_URL, 'yellow');
  
  // Check if server is running by testing stats endpoint
  log('\n1. Checking server connection...', 'blue');
  const healthCheck = await makeRequest(BASE_URL + '/api/stats');
  
  if (!healthCheck.success) {
    log('❌ Server not responding. Make sure it\'s running with "npm run dev"', 'red');
    log('   Error: ' + (healthCheck.error || 'Unknown error'), 'red');
    
    if (healthCheck.error && healthCheck.error.includes('Connection refused')) {
      log('\n💡 Environment Detection:', 'yellow');
      log('   You are running this diagnostic on your local Windows machine,', 'yellow');
      log('   but the server is running on Replit.', 'yellow');
      log('\n🔧 Choose one solution:', 'yellow');
      log('   Option A: Run this diagnostic on Replit where the server is running', 'yellow');
      log('   Option B: Start the server locally:', 'yellow');
      log('     1. Open a new terminal window', 'yellow');
      log('     2. Run: run-local.bat', 'yellow');
      log('     3. Wait for "serving on localhost:5000" message', 'yellow');
      log('     4. Then run this diagnostic again', 'yellow');
    } else if (healthCheck.data && typeof healthCheck.data === 'string' && healthCheck.data.includes('<!DOCTYPE')) {
      log('   Note: Server returned HTML instead of JSON - check if API endpoints are working', 'yellow');
    }
    return;
  }
  
  log('✅ Server is running', 'green');
  
  // Check database stats
  log('\n2. Checking database stats...', 'blue');
  const stats = await makeRequest(BASE_URL + '/api/stats');
  
  if (!stats.success) {
    log('❌ Stats API failed', 'red');
    log('   Status: ' + stats.status, 'red');
    log('   Error: ' + (stats.error || 'Unknown error'), 'red');
    log('   Response: ' + JSON.stringify(stats.data, null, 2), 'yellow');
  } else {
    log('✅ Database stats loaded', 'green');
    log('   Total combinations: ' + stats.data.totalCombinations.toLocaleString(), 'blue');
    log('   Drawn combinations: ' + stats.data.drawnCombinations, 'blue');
    log('   Never drawn: ' + stats.data.neverDrawnCombinations.toLocaleString(), 'blue');
  }
  
  // Check history
  log('\n3. Checking draw history...', 'blue');
  const history = await makeRequest(BASE_URL + '/api/history');
  
  if (!history.success) {
    log('❌ History API failed', 'red');
    log('   Status: ' + history.status, 'red');
    log('   Error: ' + (history.error || 'Unknown error'), 'red');
    log('   Response: ' + JSON.stringify(history.data, null, 2), 'yellow');
  } else {
    log('✅ Draw history loaded', 'green');
    log('   Number of draws: ' + history.data.length, 'blue');
    if (history.data.length > 0) {
      log('   Latest draw: ' + new Date(history.data[0].drawDate).toLocaleDateString(), 'blue');
      log('   Numbers: ' + history.data[0].numbers.join(', '), 'blue');
      log('   Stars: ' + history.data[0].stars.join(', '), 'blue');
    }
  }
  
  // Force initialization
  log('\n4. Force initializing data...', 'blue');
  const init = await makeRequest(BASE_URL + '/api/initialize');
  
  if (!init.success) {
    log('❌ Initialization failed', 'red');
    log('   Status: ' + init.status, 'red');
    log('   Error: ' + (init.error || 'Unknown error'), 'red');
    log('   Response: ' + JSON.stringify(init.data, null, 2), 'yellow');
  } else {
    log('✅ Data initialization completed', 'green');
    if (init.data.stats) {
      log('   Draws loaded: ' + init.data.stats.drawnCombinations, 'blue');
      log('   Prediction accuracy: ' + (init.data.stats.predictionAccuracy * 100).toFixed(1) + '%', 'blue');
    }
  }
  
  // Check analytics after initialization
  log('\n5. Checking analytics after initialization...', 'blue');
  const analytics = await makeRequest(BASE_URL + '/api/analytics/numbers');
  
  if (!analytics.success) {
    log('❌ Analytics API failed', 'red');
    log('   Status: ' + analytics.status, 'red');
    log('   Error: ' + (analytics.error || 'Unknown error'), 'red');
    log('   Response: ' + JSON.stringify(analytics.data, null, 2), 'yellow');
  } else {
    log('✅ Analytics loaded', 'green');
    if (analytics.data.hotNumbers) {
      log('   Hot numbers: ' + analytics.data.hotNumbers.slice(0, 5).map(n => n.number).join(', '), 'blue');
    }
    if (analytics.data.hotStars) {
      log('   Hot stars: ' + analytics.data.hotStars.slice(0, 3).map(s => s.number).join(', '), 'blue');
    }
  }
  
  // Final recommendation
  log('\n📋 Diagnostic Summary:', 'blue');
  
  const allWorking = healthCheck.success && stats.success && history.success && init.success && analytics.success;
  
  if (allWorking) {
    log('✅ All systems working correctly!', 'green');
    log('   Your local setup is ready to use at http://localhost:5000', 'green');
  } else {
    log('⚠️  Some issues detected. Recommendations:', 'yellow');
    
    if (!stats.success || !history.success || !analytics.success) {
      log('   1. Check your DATABASE_URL in .env file', 'yellow');
      log('   2. Run: npm run db:push', 'yellow');
      log('   3. Restart the server: npm run dev', 'yellow');
      log('   4. Visit: http://localhost:5000/api/initialize', 'yellow');
    }
  }
  
  log('\n🏁 Diagnostic complete.\n', 'blue');
}

// Run the diagnostic
diagnose().catch(console.error);
